Save mode in core

salomegeo opened the pull request:
Make mode a field on globalSettings.
Send it back to core.
Restore it from storage when app/extension is restarted.

Tested manually.




<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/uproxy/uproxy/1056)
<!-- Reviewable:end -->

lucyhe commented:
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1056)



---

Just to make sure I completely understand the distinction, to set the values we always use "ui.view =" and "ui.mode =", but to access, we always use ui.uiState.view & ui.uiState.mode, is that right?
If that's how it works, looks good to me! :+1:

---

<sup>**[src/firefox/data/scripts/firefox_browser_api.ts, line 9 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1056#-Jjb7TITqanxNxTK5YQG)** ([raw file](https://github.com/uproxy/uproxy/blob/53e94aef44185db50f7089add30e22921c206459/src/firefox/data/scripts/firefox_browser_api.ts#L9)):</sup>
just wondering, why does this change?

---


<!-- Sent from Reviewable.io -->

salomegeo commented:
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1056)



---

Correct, Is that too ugly? So initially I had getter function too for ui.view and ui.mode but for some reason it didn't work. Polymer weird binding stuff.

---

<sup>**[src/firefox/data/scripts/firefox_browser_api.ts, line 9 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1056#-Jjb7TITqanxNxTK5YQG)** ([raw file](https://github.com/uproxy/uproxy/blob/53e94aef44185db50f7089add30e22921c206459/src/firefox/data/scripts/firefox_browser_api.ts#L9)):</sup>
because this file needs ui.ts, and before it was accessing throught interfaces/ui.d.ts, but it shouldn't be. 

---


<!-- Sent from Reviewable.io -->

lucyhe commented:
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1056)

Reviewed files:
- Gruntfile.coffee @ r1
- src/chrome/extension/scripts/chrome_core_connector.ts @ r1
- src/firefox/data/scripts/firefox_browser_api.ts @ r1
- src/generic_core/core.ts @ r1
- src/generic_core/user.spec.ts @ r1
- src/generic_ui/polymer/contact.html @ r1
- src/generic_ui/polymer/contact.ts @ r1
- src/generic_ui/polymer/copypaste.ts @ r1
- src/generic_ui/polymer/instance.html @ r1
- src/generic_ui/polymer/instance.ts @ r1
- src/generic_ui/polymer/network.ts @ r1
- src/generic_ui/polymer/root.html @ r1
- src/generic_ui/polymer/root.ts @ r1
- src/generic_ui/polymer/roster.ts @ r1
- src/generic_ui/scripts/ui.spec.ts @ r1
- src/generic_ui/scripts/ui.ts @ r1
- src/generic_ui/scripts/user.ts @ r1
- src/interfaces/ui.d.ts @ r1
- src/interfaces/user.d.ts @ r1
- src/uproxy.ts @ r1

---

The only possible way I can think of getting {{ ui.view }} correctly pointing to updated values in global ui is to use a Polymer watcher (https://www.polymer-project.org/docs/polymer/polymer.html#change-watchers) on this.ui.uiState.view. When that changes, you can call a polymer function that sets this.ui.view (which won't affect global ui.view). I feel like adding an observer/update function to each Polymer element that uses ui.view is a little too wordy for just getting {{ ui.view }} over {{ ui.uiState.view }} though. Maybe best action to take is adding comments that explains the difference?

---



<!-- Sent from Reviewable.io -->

jpevarnek commented:
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1056)

Reviewed files:
- Gruntfile.coffee @ r1
- src/chrome/extension/scripts/chrome_core_connector.ts @ r1
- src/generic_core/core.ts @ r1
- src/generic_ui/polymer/contact.html @ r1
- src/generic_ui/polymer/contact.ts @ r1
- src/generic_ui/polymer/copypaste.ts @ r1
- src/generic_ui/polymer/instance.html @ r1
- src/generic_ui/polymer/instance.ts @ r1
- src/generic_ui/polymer/network.ts @ r1
- src/generic_ui/polymer/root.html @ r1
- src/generic_ui/polymer/root.ts @ r1
- src/generic_ui/polymer/roster.ts @ r1
- src/generic_ui/scripts/ui.spec.ts @ r1
- src/generic_ui/scripts/ui.ts @ r1
- src/generic_ui/scripts/user.ts @ r1
- src/interfaces/ui.d.ts @ r1
- src/interfaces/user.d.ts @ r1
- src/uproxy.ts @ r1

---

Just a couple questions, mostly on design decisions.

It would be really helpful if you could also add some text to the pull request explaining what is actually being changed here, I'm still a bit confused about the wider context of why creating the `uiState` object within core is necessary, and what it will be used for.  Also, the obligatory testing note is missing.

Regarding setting/accessing the `view` and `mode` properties, it is pretty disconcerting reading them that way.  I think it might be worth looking into alternative ideas for how to represent this (having explicit `setMode` and `setView` functions on `UiState` is the first thing that comes to mind, however, I'm still a bit fuzzy on what `UiState` is supposed to represent so that might be bad)

---

<sup>**[src/chrome/extension/scripts/chrome_core_connector.ts, line 141 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1056#-Jjeex_IcFZWodIl9oIg)** ([raw file](https://github.com/uproxy/uproxy/blob/53e94aef44185db50f7089add30e22921c206459/src/chrome/extension/scripts/chrome_core_connector.ts#L141)):</sup>
Can you fix the typo while you're in this area? (sorry, just really bothering me)

---

<sup>**[src/generic_ui/scripts/ui.ts, line 123 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1056#-JjeaOls19guFDJpH7qj)** ([raw file](https://github.com/uproxy/uproxy/blob/53e94aef44185db50f7089add30e22921c206459/src/generic_ui/scripts/ui.ts#L123)):</sup>
Can you add a comment clarifying what uiState is supposed to represent?

---

<sup>**[src/generic_ui/scripts/ui.ts, line 649 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1056#-JjeccBUbtTY8hNRv4qY)** ([raw file](https://github.com/uproxy/uproxy/blob/53e94aef44185db50f7089add30e22921c206459/src/generic_ui/scripts/ui.ts#L649)):</sup>
It seems like the normal convention for handling these events would be to add an `updateView` function to `CoreApi`, implement that function in both `CoreConnector` and `uProxyCore`.  It seems like that maintains the link between the two parts of the app a bit better than having the core commands sent from all over the app.  Is there a reason I'm missing for why that was not done in this case?

---

<sup>**[src/generic_ui/scripts/user.ts, line 13 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1056#-JjefE1HU2EVs19tvOq6)** ([raw file](https://github.com/uproxy/uproxy/blob/53e94aef44185db50f7089add30e22921c206459/src/generic_ui/scripts/user.ts#L13)):</sup>
Can you explain the reason for this move?

---

<sup>**[src/uproxy.ts, line 274 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1056#-Jjeg5ApvBGn-tamWl6d)** ([raw file](https://github.com/uproxy/uproxy/blob/53e94aef44185db50f7089add30e22921c206459/src/uproxy.ts#L274)):</sup>
I think this might look a bit better naming-wise as `UIState`.  I've never really given much thought to the question of acronyms at the beginning of camelcase variables before though...

---

<sup>**[src/uproxy.ts, line 345 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1056#-JjefJ9Vq_EKkwlpYywf)** ([raw file](https://github.com/uproxy/uproxy/blob/53e94aef44185db50f7089add30e22921c206459/src/uproxy.ts#L345)):</sup>
Extra line maybe?  I'm not sure whether this is reviewable showing the obligatory blank line at the end of the file or if this is actually an extra line though.

---


<!-- Sent from Reviewable.io -->

dborkan commented:
:+1: from me, but some things to consider as we move more UI stuff out of the background page:
* in uProxy-for-cloud, it might be possible that one uProxy core (i.e. one instance of uProxy) is running, but the user has launched multiple UIs (browser tabs).  In this case we might need to revisit how we are storing one core.uiState.view per core - I imagine each tab might be able to be on different views at a time.
* There is a bunch more UI state beyond just view and mode - pretty much everything in the ui and model and core_connector.  Some of this we might need to store in the core (ui copyAndPaste fields, splashState), other stuff we can probably just generate each time the UI launches (ui.mapInstanceIdToUser_), other stuff we might want to modify by changing the API between the core and ui (e.g. ui.instancesGivingAccessTo and ui.instanceGettingAccessFrom_ are only set when getting/sharing starts/stops, it might be better to have GET_INITIAL_STATE return these list of instances and have some other API to get them when they change).
* Moving core_connector.promiseId_ is going to take some special work - if we end up resetting the promiseId_ everytime the ui opens and closes things will break.

salomegeo commented:
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1056)



---

Changed it as discussed earlier. PTAL!

---

<sup>**[src/chrome/extension/scripts/chrome_core_connector.ts, line 141 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1056#-Jjeex_IcFZWodIl9oIg)** ([raw file](https://github.com/uproxy/uproxy/blob/53e94aef44185db50f7089add30e22921c206459/src/chrome/extension/scripts/chrome_core_connector.ts#L141)):</sup>
Done.

---

<sup>**[src/generic_ui/scripts/user.ts, line 13 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1056#-JjefE1HU2EVs19tvOq6)** ([raw file](https://github.com/uproxy/uproxy/blob/53e94aef44185db50f7089add30e22921c206459/src/generic_ui/scripts/user.ts#L13)):</sup>
Because it had a field User and I don't think it's a good idea to include .ts file from .d.ts.

This was a result of chain reaction. One thing cause another.

---

<sup>**[src/uproxy.ts, line 345 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1056#-JjefJ9Vq_EKkwlpYywf)** ([raw file](https://github.com/uproxy/uproxy/blob/53e94aef44185db50f7089add30e22921c206459/src/uproxy.ts#L345)):</sup>
Done.

---


<!-- Sent from Reviewable.io -->

jpevarnek commented:
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1056)

Reviewed files:
- src/chrome/extension/scripts/chrome_core_connector.ts @ r2
- src/generic_core/core.ts @ r3
- src/generic_core/user.spec.ts @ r1
- src/generic_ui/polymer/confirm.ts @ r2
- src/generic_ui/polymer/contact.html @ r2
- src/generic_ui/polymer/instance.ts @ r2
- src/generic_ui/polymer/root.html @ r3
- src/generic_ui/polymer/root.ts @ r2
- src/generic_ui/scripts/ui.spec.ts @ r2
- src/generic_ui/scripts/ui.ts @ r2
- src/interfaces/persistent.d.ts @ r2
- src/uproxy.ts @ r2





---

<sup>**[src/generic_ui/scripts/ui.spec.ts, line 321 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/1056#-JjjtR77ri1IIv7sZET2)** ([raw file](https://github.com/uproxy/uproxy/blob/16e9aaf133337e869b4112c4ebc60f25cbb572f9/src/generic_ui/scripts/ui.spec.ts#L321)):</sup>
Can this just be completely removed?

---

<sup>**[src/generic_ui/scripts/ui.ts, line 42 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/1056#-JjjtX8mNGVYM_KebP2J)** ([raw file](https://github.com/uproxy/uproxy/blob/16e9aaf133337e869b4112c4ebc60f25cbb572f9/src/generic_ui/scripts/ui.ts#L42)):</sup>
Quotes are unnecessary for the attribute name here

---

<sup>**[src/generic_ui/scripts/ui.ts, line 124 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/1056#-JjjtdsHONapRM9Qzbmk)** ([raw file](https://github.com/uproxy/uproxy/blob/16e9aaf133337e869b4112c4ebc60f25cbb572f9/src/generic_ui/scripts/ui.ts#L124)):</sup>
"Appearance." does not really add anything to the name of the variable in this case

---

<sup>**[src/generic_ui/scripts/user.ts, line 13 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1056#-JjefE1HU2EVs19tvOq6)** ([raw file](https://github.com/uproxy/uproxy/blob/53e94aef44185db50f7089add30e22921c206459/src/generic_ui/scripts/user.ts#L13)):</sup>
Grepping the code real quick, it looks like this interface might not even be used anywhere.  If so, can you just pull it out?

---


<!-- Sent from Reviewable.io -->

salomegeo commented:
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1056)



---

PTAL

---

<sup>**[src/generic_ui/scripts/ui.spec.ts, line 321 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/1056#-JjjtR77ri1IIv7sZET2)** ([raw file](https://github.com/uproxy/uproxy/blob/16e9aaf133337e869b4112c4ebc60f25cbb572f9/src/generic_ui/scripts/ui.spec.ts#L321)):</sup>
Done.

---

<sup>**[src/generic_ui/scripts/ui.ts, line 42 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/1056#-JjjtX8mNGVYM_KebP2J)** ([raw file](https://github.com/uproxy/uproxy/blob/16e9aaf133337e869b4112c4ebc60f25cbb572f9/src/generic_ui/scripts/ui.ts#L42)):</sup>
Done.

---

<sup>**[src/generic_ui/scripts/ui.ts, line 124 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/1056#-JjjtdsHONapRM9Qzbmk)** ([raw file](https://github.com/uproxy/uproxy/blob/16e9aaf133337e869b4112c4ebc60f25cbb572f9/src/generic_ui/scripts/ui.ts#L124)):</sup>
Do you want me to remove it?

---

<sup>**[src/generic_ui/scripts/user.ts, line 13 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1056#-JjefE1HU2EVs19tvOq6)** ([raw file](https://github.com/uproxy/uproxy/blob/53e94aef44185db50f7089add30e22921c206459/src/generic_ui/scripts/user.ts#L13)):</sup>
Done.

---


<!-- Sent from Reviewable.io -->

jpevarnek commented:
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1056)







---

<sup>**[src/generic_ui/scripts/ui.ts, line 124 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/1056#-JjjtdsHONapRM9Qzbmk)** ([raw file](https://github.com/uproxy/uproxy/blob/16e9aaf133337e869b4112c4ebc60f25cbb572f9/src/generic_ui/scripts/ui.ts#L124)):</sup>
I'd say so but don't feel strongly about it.  Either remove it or add an explanation that it's which view we are currently showing in the UI

---


<!-- Sent from Reviewable.io -->

salomegeo commented:
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1056)







---

<sup>**[src/generic_ui/scripts/ui.ts, line 124 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/1056#-JjjtdsHONapRM9Qzbmk)** ([raw file](https://github.com/uproxy/uproxy/blob/16e9aaf133337e869b4112c4ebc60f25cbb572f9/src/generic_ui/scripts/ui.ts#L124)):</sup>
FYI that comment was there.

As long as you don't ask me to add more comment, I'm happy to delete anything.

---


<!-- Sent from Reviewable.io -->

salomegeo commented:
@dborkan can you take a look? I want to get this submitted and merged in my current branch so that I can finally finish this background page work.

dborkan commented:
:+1: but can you update the description of this pull request with more details about why you are making these changes and how they were tested?  Thanks

salomegeo commented:
@jpevarnek this pull request pretty much completely changed, but I addressed all your comments that were still relevant.

Can you please take a look if there's anything else?

jpevarnek commented:
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1056)

Reviewed files:
- src/generic_ui/polymer/instance.html @ r2
- src/generic_ui/polymer/root.html @ r4
- src/generic_ui/polymer/root.ts @ r4
- src/generic_ui/scripts/ui.spec.ts @ r4
- src/generic_ui/scripts/ui.ts @ r5
- src/generic_ui/scripts/user.ts @ r4





---

<sup>**[src/generic_ui/polymer/root.html, line 177 \[r5\]](https://reviewable.io:443/reviews/uproxy/uproxy/1056#-Jk3zMVwXdey2vR6yME4)** ([raw file](https://github.com/uproxy/uproxy/blob/5d530060d6d2ffab8e88bb875e55a13781db10ad/src/generic_ui/polymer/root.html#L177)):</sup>
This ID feels a bit wrong (it's referring to a single tab while on the tab bar element)

---


<!-- Sent from Reviewable.io -->

jpevarnek commented:
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1056)



---

Minor nitpick, otherwise LGTM

---



<!-- Sent from Reviewable.io -->

salomegeo commented:
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1056)







---

<sup>**[src/generic_ui/polymer/root.html, line 177 \[r5\]](https://reviewable.io:443/reviews/uproxy/uproxy/1056#-Jk3zMVwXdey2vR6yME4)** ([raw file](https://github.com/uproxy/uproxy/blob/5d530060d6d2ffab8e88bb875e55a13781db10ad/src/generic_ui/polymer/root.html#L177)):</sup>
Done.

---


<!-- Sent from Reviewable.io -->

jpevarnek commented:
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1056)



---

LGTM

---



<!-- Sent from Reviewable.io -->

jpevarnek commented:
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1056)

Reviewed files:
- src/generic_ui/polymer/root.html @ r6
- src/generic_ui/polymer/root.ts @ r6





---



<!-- Sent from Reviewable.io -->
